ratis-client/src/main/java/org/apache/ratis/client/DataStreamClient.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ratis.client;
+
+import org.apache.ratis.RaftConfigKeys;
+import org.apache.ratis.datastream.DataStreamApi;
+import org.apache.ratis.client.impl.DataStreamClientImpl;
+import org.apache.ratis.conf.Parameters;
+import org.apache.ratis.conf.RaftProperties;
+import org.apache.ratis.datastream.SupportedDataStreamType;
+import org.apache.ratis.datastream.objects.DataStreamReply;
+import org.apache.ratis.datastream.objects.DataStreamRequest;
+import org.apache.ratis.protocol.ClientId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A client interface that sends request to the streaming pipeline.
+ * Associated with it will be a Netty Client.
+ */
+public interface DataStreamClient {
+
+  Logger LOG = LoggerFactory.getLogger(DataStreamClient.class);
+
+  /** Return Client id. */
+  ClientId getId();
+
+  /** Return Streamer Api instance. */
+  DataStreamApi getDataStreamApi();
+
+  /** Return Network transfer Api instance. */
+  RaftClientStream getNetworkTransferApi();
+
+  /**
+   * send to server via streaming.
+   * Return a completable future.
+   */
+  CompletableFuture<DataStreamReply> sendAsync(DataStreamRequest request);
```

Comment:
This method also can be removed.  We expect the user application to only use DataStreamApi create a stream and then send data using the stream.

Reply:
We would need this method right? Since the raftClientStream is attached to DataStreamClient instead of DataStreamApi. @szetszwo

Reply:
User application call RaftDataOutputStream.sendAsync(..).  Then, internally, the RaftDataOutputStream impl will call RaftClientStream.sendRequestAsync(..) to send it out to network.

Who and when will call DataStreamClient.sendAsync?

Reply:
oh sorry! understood. I was just thinking of adding a layer between RaftDataOutputStream.sendAsync(..) -> DataStreamClient.sendAsync() -> RaftClientStream.sendRequestAsync(..). We can remove that.